wai-handler-launch/Network/Wai/Handler/Launch.hs

```diff
@@ -192,20 +193,36 @@ runUrlPort = runHostPortUrl "*4"
 
 runHostPortUrl :: String -> Int -> String -> Application -> IO ()
 runHostPortUrl host port url app = do
-    x <- newIORef True
-    _ <- forkIO $ Warp.runSettings
-        ( Warp.setPort port
-        $ Warp.setOnException (\_ _ -> return ())
-        $ Warp.setHost (fromString host) Warp.defaultSettings)
-        $ ping x app
-    launch port url
-    loop x
+    ready <- newEmptyMVar
+    active <- newIORef True
+    let settings =
+          Warp.setPort port $
+          Warp.setOnException (\_ _ -> return ()) $
+          Warp.setHost (fromString host) $
+          Warp.setBeforeMainLoop (putMVar ready ()) $
+          Warp.defaultSettings
+    withAsync
```

Comment:
Sorry to do these piece-meal, but I only notice these things after each successive issue has been addressed.

I think this `withAsync`+`link` is identical in behavior to `fmap (either id id) $ race (Warp.runSettings ...) (takeMVar ready ...)`